x/concentrated-liquidity/store.go

```diff
@@ -16,10 +19,85 @@ func (k Keeper) getAllPositionsWithVaryingFreezeTimes(ctx sdk.Context, poolId ui
 }
 
+// ParsePositionFromBz parses a position from a byte array.
+// Returns a struct containing the liquidity associated with the position.
+// Returns an error if the byte array is empty.
+// Returns an error if fails to parse.
 func ParsePositionFromBz(bz []byte) (position model.Position, err error) {
 	if len(bz) == 0 {
 		return model.Position{}, errors.New("position not found")
 	}
 	err = proto.Unmarshal(bz, &position)
 	return position, err
 }
+
+// ParseFullPositionFromBytes parses a full position from key and value bytes.
+// Returns a struct containing the pool id, lower tick, upper tick, frozen until, and liquidity
+// associated with the position.
+// Returns an error if the key or value is not found.
+// Returns an error if fails to parse either.
+func ParseFullPositionFromBytes(key, value []byte) (types.FullPositionByOwnerResult, error) {
+	if len(key) == 0 {
+		return types.FullPositionByOwnerResult{}, errors.New("key not found")
+	}
+	if len(value) == 0 {
+		return types.FullPositionByOwnerResult{}, fmt.Errorf("value not found for key (%s)", value)
+	}
+
+	keyStr := string(key)
+
+	// These may include irrelevant parts of the prefix such as the module prefix
+	// and position prefix.
+	fullPositionKeyComponents := strings.Split(keyStr, types.KeySeparator)
+
+	if len(fullPositionKeyComponents) < 6 {
+		return types.FullPositionByOwnerResult{}, fmt.Errorf(`invalid position key (%s), must have at least 5 components:
+	(position prefix, owner address, pool id, lower tick, upper tick, frozen until),
+	all separated by (%s)`, keyStr, types.KeySeparator)
+	}
+
+	// We only care about the last 5 components, which are:
+	// - pool id
+	// - lower tick
+	// - upper tick
+	// - frozen until
+	relevantPositionKeyComponents := fullPositionKeyComponents[len(fullPositionKeyComponents)-4:]
+
+	positionPrefix := fullPositionKeyComponents[0]
+	if positionPrefix != string(types.PositionPrefix) {
+		return types.FullPositionByOwnerResult{}, fmt.Errorf("Wrong position prefix, got: %v, required %v", []byte(positionPrefix), types.PositionPrefix)
+	}
+
+	poolId, err := strconv.ParseUint(relevantPositionKeyComponents[0], 10, 64)
+	if err != nil {
+		return types.FullPositionByOwnerResult{}, err
+	}
+
+	lowerTick, err := strconv.ParseInt(relevantPositionKeyComponents[1], 10, 64)
+	if err != nil {
+		return types.FullPositionByOwnerResult{}, err
+	}
+
+	upperTick, err := strconv.ParseInt(relevantPositionKeyComponents[2], 10, 64)
+	if err != nil {
+		return types.FullPositionByOwnerResult{}, err
+	}
+
+	frozenUntil, err := osmoutils.ParseTimeString(relevantPositionKeyComponents[3])
+	if err != nil {
+		return types.FullPositionByOwnerResult{}, err
+	}
+
+	positionValue, err := ParsePositionFromBz(value)
+	if err != nil {
+		return types.FullPositionByOwnerResult{}, err
+	}
+
+	return types.FullPositionByOwnerResult{
+		PoolId:      poolId,
+		LowerTick:   lowerTick,
+		UpperTick:   upperTick,
+		FrozenUntil: frozenUntil,
```

Comment:
What do you think about adding a check that `frozenUntil` which we get from the key is equal to `frozenUntil` in `positionValue`? Just a sanity check. I hesitate is we should add it, because it sounds a little bit redundant

@p0mvn

Reply:
I think we only have liquidity in the value, there should be no `frozenUntil` as far as I know

Reply:
yeah, my idea was to, like, make sure the position in the value corresponds with the key. And by comparing `frozenUntil` in both at least, we know up to some degree that it is the correct value.

Reply:
and because if `frozenUntil`s are not the same in key and value, we can't be sure that the liquidity is correct in the value 
The reason I thought it is redundant: if a store provides a guarantee that there is absolutely no way for key with lets say `frozenUntil = 5` to return `frozenUntil = 10` value, then yes my suggestion is redundant

Reply:
I hope that makes sense

Reply:
Ah I see. I think we do not need to have that in both key and value.

I think we should merge this PR as well but investigate separately.

Having `frozenUntil` in both key and value is redundant. If we use it in key, we can also extract it from key.

CC: @czarcas7ic @AlpinYukseloglu